--- a/c/lib.c
+++ b/c/lib.c
@@ -1,13 +1,15 @@
 #include <radamsa.h>

 void radamsa_init() {
+   if (memstart != NULL)
+      free(memstart);
    int nobjs=0, nwords=0;
    hp = (byte *) &heap; /* builtin heap */
    state = IFALSE;
    heap_metrics(&nwords, &nobjs);
    max_heap_mb = (W == 4) ? 4096 : 65535;
    nwords += nobjs + INITCELLS;
-   memstart = genstart = fp = (word *) realloc(NULL, (nwords + MEMPAD)*W);
+   memstart = genstart = fp = (word *) malloc((nwords + MEMPAD)*W);
    if (!memstart) return;
    memend = memstart + nwords - MEMPAD;
    state = (word) load_heap(nobjs);
